### PR TITLE
Support attention data parallelism

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

This work was done in collaboration with @richjames0 and @NicoGrande

This PR introduces attention data parallelism (attn_dp) to optimize memory efficiency when the number of KV heads is less than tensor parallelism. The attention DP degree is auto-calculated based on the ratio of tensor parallelism to KV heads, ensuring optimal sharding without manual configuration.
New logical axes (attn_activation_length, attn_activation_embed) and corresponding sharding rules have been added to support attention-specific tensor partitioning separate from the rest of the model

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Ran Llama3.1-8B by reducing num_kv_heads to 2 and ici_tensor_parallelism=8. This will auto-calculate attn_dp and set it to 4. 
vllm mesh: `mesh=Mesh('data': 1, 'attn_dp': 4, 'expert': 1, 'model': 2, axis_types=(Auto, Auto, Auto, Auto))`

Attention shardings:
```
llama2.py:158] bfloat16[96,1,4096]............................................................. (None, None, ('model', 'attn_dp')).
attentions.py:1014] bfloat16[96,1,4096]............................................................. (None, 'attn_dp', 'model').
attentions.py:1078] bfloat16[96,1,32,128]........................................................... (None, 'attn_dp', 'model', None).
attentions.py:1079] bfloat16[96,1,8,128]............................................................ (None, 'attn_dp', 'model', None).
linears.py:527] bfloat16[96,1,14336]............................................................ (None, None, ('model', 'attn_dp')).
```

Weight shardings:
```
maxtext_utils.py:1197] decoder/decoder_norm/scale/value................................................ bfloat16[4096] (('model', 'attn_dp'),)
maxtext_utils.py:1197] decoder/layers_0/mlp/wi_0/kernel/value.......................................... bfloat16[4096,14336] (None, ('model', 'attn_dp'))
maxtext_utils.py:1197] decoder/layers_0/mlp/wi_1/kernel/value.......................................... bfloat16[4096,14336] (None, ('model', 'attn_dp'))
maxtext_utils.py:1197] decoder/layers_0/mlp/wo/kernel/value............................................ bfloat16[14336,4096] (('model', 'attn_dp'), None)
maxtext_utils.py:1197] decoder/layers_0/post_self_attention_layer_norm/scale/value..................... bfloat16[4096] (('model', 'attn_dp'),)
maxtext_utils.py:1197] decoder/layers_0/pre_self_attention_layer_norm/scale/value...................... bfloat16[4096] (('model', 'attn_dp'),)
maxtext_utils.py:1197] decoder/layers_0/self_attention/key/kernel/value................................ bfloat16[4096,8,128] (None, 'model', None)
maxtext_utils.py:1197] decoder/layers_0/self_attention/out/kernel/value................................ bfloat16[32,128,4096] ('model', None, None)
maxtext_utils.py:1197] decoder/layers_0/self_attention/query/kernel/value.............................. bfloat16[4096,32,128] (None, 'model', None)
maxtext_utils.py:1197] decoder/layers_0/self_attention/value/kernel/value.............................. bfloat16[4096,8,128] (None, 'model', None)
decoder/logits_dense/kernel/value............................................... bfloat16[4096,128256] (None, ('model', 'attn_dp'))
token_embedder/embedding/value.................................................. bfloat16[128256,4096] (('model', 'attn_dp'), None)
```

Running Qwen3-30b-moe with tp=8, it has 4 kv heads so model=4 and attn_dp=2
Command: 
```
NEW_MODEL_DESIGN=1 python3 -m MaxText.vllm_decode     --model_name qwen3-30b-a3b  --hf_model_name Qwen/Qwen3-30B-A3B     --hf_config_path src/MaxText/integration/vllm/maxtext_vllm_adapter  --ici_tensor_parallelism 8  --gpu_memory_utilization 0.5  --prompt "Suggest some famous landmarks in London" --debug_sharding true --enable_dp_attention true --load_parameters_path gs://parambole-qwen3-moe-verification/unscanned/qwen3-30b-a3b-thinking-2507/14_08_2025/0/items 2>&1 | tee attn_output
```
Output: https://paste.googleplex.com/5980793354715136

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/configs/base.yml [hunk 1]
```diff
@@ -383,7 +383,7 @@ hardware: 'tpu' # Supported hardware types are 'tpu', 'gpu', 'gpu_multiprocess'
 
 # Parallelism
 shard_mode: "auto" # can be either auto or explicit
-mesh_axes: ['data', 'stage', 'fsdp', 'fsdp_transpose', 'sequence', 'context', 'context_autoregressive', 'tensor', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive']
+mesh_axes: ['data', 'stage', 'attn_dp', 'fsdp', 'fsdp_transpose', 'sequence', 'context', 'context_autoregressive', 'tensor', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive']
```
Comment (NicoGrande) [inline src/MaxText/configs/base.yml]: I don't think we need `attn_dp` here since we are not using it for training.

File: src/MaxText/configs/types.py [hunk 2]
```diff
@@ -2174,12 +2176,56 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
...
+        self.ici_tensor_parallelism = num_kv_heads
+      else:
+        logger.info(
+            "Attention DP not needed: num_kv_heads=%s >= ici_tensor_parallelism=%s. "
+            "Keeping ici_attn_dp_parallelism=1.",
+            num_kv_heads,
+            tp_size,
+        )
+        self.ici_attn_dp_parallelism = 1
+
```
Comment (NicoGrande) [inline src/MaxText/configs/types.py]: This logic sets the `ici_attn_dp_parallelism` similar to how its done in `tpu-inference`. Since the mesh itself is being initialized in `tpu-inference` do we need this code in MaxText? I don't think we will have `attn_dp` in non-vLLM code paths right?

File: src/MaxText/layers/moe.py [hunk 3]
```diff
@@ -355,7 +355,7 @@ def __init__(
 
     if self.config.attention == "vllm_rpa":
       # vLLM uses 'model' as the tensor parallelism axis name
-      self._tensor_parallelism_name = "model"
+      self._tensor_parallelism_name = ("model", "attn_dp")
```
Comment (NicoGrande) [inline src/MaxText/layers/moe.py:358]: Does this also work for the case where `attn_dp` is 1?

File: src/MaxText/configs/types.py (hunk 2 above)
Comment (NicoGrande) [reply src/MaxText/configs/types.py]: I think this is also true of other changes in `types.py`

File: src/MaxText/configs/base.yml (hunk 1 above)
Comment (khatwanimohit) [reply src/MaxText/configs/base.yml]: removed this

File: src/MaxText/configs/types.py (hunk 2 above)
Comment (khatwanimohit) [reply src/MaxText/configs/types.py]: removed this part, just kept `attn_dp` in ici/dcn_map to get around pyconfig initialization with vllm mesh axes

File: src/MaxText/layers/moe.py (hunk 3 above)
Comment (khatwanimohit) [reply src/MaxText/layers/moe.py:358]: yes confirmed things are working with tp=4 and attn_dp=1 https://paste.googleplex.com/4958033555881984

Comment (NicoGrande) [reply src/MaxText/layers/moe.py:358]: Thank you!

**Review by NicoGrande — APPROVED** [review]
LGTM! Awesome work!

File: src/MaxText/configs/base.yml [hunk 4]
```diff
@@ -393,6 +393,10 @@ logical_axis_rules: [
                       ['activation_kv_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence']],
                       ['activation_length', ['sequence', 'context', 'expert']],
                       ['activation_length', ['context', 'expert']],
+                      ['attn_activation_length', ['sequence', 'context', 'expert']],
+                      ['attn_activation_length', ['context', 'expert']],
+                      ['attn_activation_length_no_exp', ['sequence', 'context']],
+                      ['attn_activation_length_no_exp', ['context']],
```
Comment (NuojCheng) [inline src/MaxText/configs/base.yml]: I would prefer `activation_attn_length` with 'activation' ahead to emphasize it is for activation

File: src/MaxText/configs/base.yml [hunk 5]
```diff
@@ -401,6 +405,7 @@ logical_axis_rules: [
                       ['prefill_activation_length', ['sequence', 'context']],
                       ['prefill_activation_norm_length', ['tensor_sequence', 'context', 'sequence']],
                       ['activation_kv_length', []],
+                      ['attn_activation_embed', ['tensor', 'tensor_transpose']],
```
Comment (NuojCheng) [inline src/MaxText/configs/base.yml]: do we plan to use a separate PR to moe part? This change won't enable full EP sharding on MOE.

File: src/MaxText/configs/types.py [hunk 6]
```diff
@@ -2220,6 +2220,7 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
           "model": self.ici_tensor_parallelism,
           "expert": self.ici_expert_parallelism,
           "autoregressive": self.ici_autoregressive_parallelism,
+          "attn_dp": 1,  # initialized to 1, vLLM will auto calculate this value based on TP and num_kv_heads
```
Comment (NuojCheng) [inline src/MaxText/configs/types.py:2223]: do we have to make this change? in vllm.yml it uses "model" mesh axis but it is not specified here. I think "model" and "attn_dp" play similar roles?

Comment (khatwanimohit) [reply src/MaxText/configs/types.py:2223]: I also didn't want this change, but this change is needed just so the pyconfig is initialized correctly with vllm.yml's mesh axes. In actuality the mesh is created by vllm and attn_dp is set by vllm based on the values of TP and num_kv_heads

File: src/MaxText/configs/base.yml (hunk 5 above)
Comment (khatwanimohit) [reply src/MaxText/configs/base.yml]: we haven't verified these logical axis rules with EP sharding yet, that's the plan in an upcoming PR

File: src/MaxText/configs/base.yml (hunk 4 above)
Comment (NuojCheng) [reply src/MaxText/configs/base.yml]: we probably should have `activation_attn_batch` as well, unless you want `attn_dp` plays as TP instead of DP in attention component.

File: src/MaxText/configs/vllm.yml [hunk 7]
```diff
@@ -25,43 +25,49 @@ weight_dtype: bfloat16
...
                       ['kv_heads', ['model']],
                       ['kv_head_dim', []],
                       ['kv', []],
                       ['embed', ['expert']],
+                      ['embed_tensor_transpose', ['attn_dp', 'model']],
                       ['embed_no_exp', []],
                       ['q_lora', ['expert']],
                       ['kv_lora', ['expert']],
-                      ['norm', ['model']],
+                      ['norm', []],
```
Comment (NuojCheng) [inline src/MaxText/configs/vllm.yml:67]: just for my own understanding, why remove TP axis for `norm`?

File: src/MaxText/configs/types.py (hunk 6 above)
Comment (NuojCheng) [reply src/MaxText/configs/types.py:2223]: after reading moe.py change, now I understand why. please ignore

File: src/MaxText/maxtext_utils.py [hunk 8]
```diff
@@ -1184,3 +1184,15 @@ def print_state_mesh_shardings_params(state, state_sharding, mesh):
...
+def print_shardings_params(params, params_sharding, mesh):
+  """Print params shardings."""
+  leaves_params, _ = jax.tree_util.tree_flatten_with_path(params)
+  leaves_sharding, _ = jax.tree_util.tree_flatten_with_path(params_sharding)
+
+  for (path, leaf_val), (_, leaf_sharding) in zip(leaves_params, leaves_sharding):
+    path_str = "/".join(str(p.key) for p in path[:-1]) + "/" + path[-1].name
+    shape = jax.typeof(leaf_val)
+    pspec = sharding.remove_size_one_mesh_axis(leaf_sharding.spec, mesh)
+    max_logging.log(f"{path_str:.<80} {shape} {tuple(pspec)}")
```
Comment (NuojCheng) [inline src/MaxText/maxtext_utils.py:1186]: Is it possible to merge this function with the function above [heart:]

File: src/MaxText/model_creation_utils.py [hunk 9]
```diff
@@ -153,7 +154,10 @@ def create_sharded_state():
     with nn.logical_axis_rules(config.logical_axis_rules):
       sharded_state = create_sharded_state()
     model = nnx.merge(graphdef, sharded_state)
-
+    # print weights sharding info under debug sharding mode
+    if config.debug_sharding:
+      max_utils.print_non_trivial_mesh_axis(model.mesh)
+      maxtext_utils.print_shardings_params(sharded_state, out_shardings, model.mesh)
```
Comment (NuojCheng) [inline src/MaxText/model_creation_utils.py:160]: good job!

File: src/MaxText/configs/vllm.yml (hunk 7 above)
Comment (khatwanimohit) [reply src/MaxText/configs/vllm.yml:67]: We saw that vllm is replicating the norm layer for all their models

File: src/MaxText/maxtext_utils.py (hunk 8 above)
Comment (khatwanimohit) [reply src/MaxText/maxtext_utils.py:1186]: is it okay if I change the above function to accept params instead of state ?

**Comment by khatwanimohit** [comment]
> It is surprising to see
> 
> ```
> attentions.py:1014] bfloat16[96,1,4096]............................................................. (None, 'attn_dp', 'model').
> ```
> 
> with `'attn_dp' = 4`. It should produce error instead. Is it a typo?

I just re-ran Llama3.1-8B with 2 kv heads and here are the [updated shardings](https://paste.googleplex.com/5860195102228480). You are right, this should have errored out, not sure what happened earlier. I remember in that run earlier we were missing a env variable (`NEW_MODE_DESIGN=1`)in vLLM which was responsible for using new shardings which includes attn_dp

File: src/MaxText/maxtext_utils.py (hunk 8 above)
Comment (khatwanimohit) [reply src/MaxText/maxtext_utils.py:1186]: I have now modified the function to accept params, params_sharding instead of state,state_mesh_shardings). I did this modification so I can use it in the create_nnx_model codepath

File: src/MaxText/configs/base.yml (hunk 4 above)
Comment (khatwanimohit) [reply src/MaxText/configs/base.yml]: changed the name to `activation_attn_length`. Discussed offline that activations inside attention are replicated accross `attn_dp`. `activation_attn_batch` is not added because batch size is always 1 while running forward pass with vLLM
